sycl/test-e2e/VirtualFunctions/misc/group-barrier.cpp

```diff
@@ -0,0 +1,140 @@
+// FIXME: replace unsupported with an aspect check once we have it
+// UNSUPPORTED: cuda, hip, acc
+//
+// REQUIRES: aspect-usm_shared_allocations
+//
+// Fails with UR_RESULT_ERROR_PROGRAM_LINK_FAILURE. SPIR-V files produced by
+// SYCL_DUMP_IMAGES can be linked just fine (using llvm-spirv -r + llvm-link),
+// so it seems to be a problem on IGC side.
+// Reported in https://github.com/intel/llvm/issues/15068
+// On CPU it segfaults within the kernel that performs virtual function call.
+// https://github.com/intel/llvm/issues/15080
+// XFAIL: gpu, cpu
+//
+// This test checks that group operations (barrier in this case) work correctly
+// inside virtual functions.
+//
+// RUN: %{build} -o %t.out %helper-includes
+// RUN: %{run} %t.out
+
+#include <sycl/detail/core.hpp>
+#include <sycl/group_barrier.hpp>
+#include <sycl/usm.hpp>
+
+#include "helpers.hpp"
+
+#include <iostream>
+#include <numeric>
+
+namespace oneapi = sycl::ext::oneapi::experimental;
+
+class BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *, sycl::nd_item<1>) = 0;
+};
+
+class SumOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *LocalData, sycl::nd_item<1> It) {
+    LocalData[It.get_local_id()] += It.get_local_id();
+    sycl::group_barrier(It.get_group());
+    int Res = 0;
+    if (It.get_group().leader()) {
+      for (size_t I = 0; I < It.get_local_range().size(); ++I) {
+        Res += LocalData[I];
+      }
+    }
+
+    return Res;
+  }
+};
+
+class MultiplyOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *LocalData, sycl::nd_item<1> It) {
+    LocalData[It.get_local_id()] += It.get_local_id();
+    sycl::group_barrier(It.get_group());
+    int Res = 1;
+    if (It.get_group().leader()) {
+      for (size_t I = 0; I < It.get_local_range().size(); ++I) {
+        Res *= LocalData[I];
+      }
+    }
+
+    return Res;
+  }
+};
+
+int main() try {
+  using storage_t = obj_storage_t<SumOp, MultiplyOp>;
+
+  auto asyncHandler = [](sycl::exception_list list) {
+    for (auto &e : list)
+      std::rethrow_exception(e);
+  };
```

Comment:
In my first PR with virtual function tests I've discovered that they were silently failing/crashing on Windows because my kernels used `double` (inadvertently) and my conclusion was that our default async handler on Windows is too bad to catch that.

Doing some experiments now and looking into the code, I see that our default async handler implementation is good enough and what's important is a top-level `try...catch` to also intercept synchronious exceptions.

I will remove async handler to simplify those tests

Reply:
Removed in 90c917550c30b48999ef30ef12c3ad431f1cd1d5